lib/SourceMapDevToolPlugin.js

```diff
@@ -129,25 +130,26 @@ SourceMapDevToolPlugin.prototype.apply = function(compiler) {
 					currentSourceMappingURLComment = currentSourceMappingURLComment.replace(/^\n\/\/(.*)$/, "\n/*$1*/");
 				}
 				if(sourceMapFilename) {
-					var filename = file,
-						query = "";
-					var idx = filename.indexOf("?");
-					if(idx >= 0) {
-						query = filename.substr(idx);
-						filename = filename.substr(0, idx);
+					var filename = splitQuery(file);
+					var contentHash;
+					if(sourceMapFilename.indexOf("[hash]")) {
+						var hash = crypto.createHash("md5");
+						hash.update(source);
+						contentHash = hash.digest("hex");
 					}
-					var sourceMapFile = this.getPath(sourceMapFilename, {
+					var sourceMapFile = splitQuery(this.getPath(sourceMapFilename, {
 						chunk: chunk,
-						filename: filename,
-						query: query,
-						basename: basename(filename)
-					});
-					var sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						filename: filename.name,
+						query: filename.query,
+						hash: contentHash,
+						basename: basename(filename.name)
+					}));
+					var sourceMapUrl = path.relative(path.dirname(file), sourceMapFile.name).replace(/\\/g, "/") + sourceMapFile.query;
 					if(currentSourceMappingURLComment !== false) {
 						asset.__SourceMapDevToolData[file] = this.assets[file] = new ConcatSource(new RawSource(source), currentSourceMappingURLComment.replace(/\[url\]/g, sourceMapUrl));
 					}
-					asset.__SourceMapDevToolData[sourceMapFile] = this.assets[sourceMapFile] = new RawSource(JSON.stringify(sourceMap));
-					chunk.files.push(sourceMapFile);
+					asset.__SourceMapDevToolData[sourceMapFile.name] = this.assets[sourceMapFile.name] = new RawSource(JSON.stringify(sourceMap));
+					chunk.files.push(sourceMapFile.name);
```

Comment:
about that... does it make sense that query string is part of the asset name, since it's not part of the file name?